### PR TITLE
[SPARK-22883][ML][TEST] Streaming tests for spark.ml.feature, from A to H

### PR DESCRIPTION
**Pull request by jkbradley**
## What changes were proposed in this pull request?

Adds structured streaming tests using testTransformer for these suites:
* BinarizerSuite
* BucketedRandomProjectionLSHSuite
* BucketizerSuite
* ChiSqSelectorSuite
* CountVectorizerSuite
* DCTSuite.scala
* ElementwiseProductSuite
* FeatureHasherSuite
* HashingTFSuite

## How was this patch tested?

It tests itself because it is a bunch of tests!

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketedRandomProjectionLSHSuite.scala [hunk 1]
```diff
@@ -98,6 +97,21 @@ class BucketedRandomProjectionLSHSuite
     MLTestingUtils.checkCopyAndUids(brp, brpModel)
   }
 
+  test("BucketedRandomProjectionLSH: streaming transform") {
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/BucketedRandomProjectionLSHSuite.scala]: No existing test to use

File: mllib/src/test/scala/org/apache/spark/ml/feature/ChiSqSelectorSuite.scala [hunk 2]
```diff
@@ -163,18 +162,19 @@ class ChiSqSelectorSuite extends SparkFunSuite with MLlibTestSparkContext
...
   }
-}
 
-object ChiSqSelectorSuite {
-
-  private def testSelector(selector: ChiSqSelector, dataset: Dataset[_]): ChiSqSelectorModel = {
-    val selectorModel = selector.fit(dataset)
-    selectorModel.transform(dataset).select("filtered", "topFeature").collect()
-      .foreach { case Row(vec1: Vector, vec2: Vector) =>
+  private def testSelector(selector: ChiSqSelector, data: Dataset[_]): ChiSqSelectorModel = {
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/ChiSqSelectorSuite.scala]: Moved from object to class b/c this needed testTransformer from the MLTest mix-in

File: mllib/src/test/scala/org/apache/spark/ml/feature/ElementwiseProductSuite.scala [hunk 3]
```diff
@@ -17,13 +17,31 @@
...
+import org.apache.spark.ml.util.TestingUtils._
+import org.apache.spark.sql.Row
 
-class ElementwiseProductSuite
-  extends SparkFunSuite with MLlibTestSparkContext with DefaultReadWriteTest {
+class ElementwiseProductSuite extends MLTest with DefaultReadWriteTest {
+
+  import testImplicits._
+
+  test("streaming transform") {
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/ElementwiseProductSuite.scala]: No existing unit test to use

File: mllib/src/test/scala/org/apache/spark/ml/feature/FeatureHasherSuite.scala [hunk 4]
```diff
@@ -17,26 +17,23 @@
...
-  with MLlibTestSparkContext
-  with DefaultReadWriteTest {
+class FeatureHasherSuite extends MLTest with DefaultReadWriteTest {
 
   import testImplicits._
 
   import HashingTFSuite.murmur3FeatureIdx
 
-  implicit private val vectorEncoder = ExpressionEncoder[Vector]()
+  implicit private val vectorEncoder: ExpressionEncoder[Vector] = ExpressionEncoder[Vector]()
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/FeatureHasherSuite.scala]: scala style

File: mllib/src/test/scala/org/apache/spark/ml/feature/FeatureHasherSuite.scala [hunk 5]
```diff
@@ -51,31 +48,31 @@ class FeatureHasherSuite extends SparkFunSuite
   }
 
   test("feature hashing") {
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/FeatureHasherSuite.scala]: Rearranged this test so it checks each row independently.

File: mllib/src/test/scala/org/apache/spark/ml/feature/HashingTFSuite.scala [hunk 6]
```diff
@@ -37,21 +36,28 @@ class HashingTFSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
   }
 
   test("hashingTF") {
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/HashingTFSuite.scala]: ditto: rearranged to do validity check per-row

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketedRandomProjectionLSHSuite.scala [hunk 7]
```diff
@@ -98,6 +97,21 @@ class BucketedRandomProjectionLSHSuite
...
+    val brp = new BucketedRandomProjectionLSH()
+      .setNumHashTables(2)
+      .setInputCol("keys")
+      .setOutputCol("values")
+      .setBucketLength(1.0)
+      .setSeed(12345)
+    val brpModel = brp.fit(dataset)
+
+    testTransformer[Tuple1[Vector]](dataset.toDF(), brpModel, "keys", "values") {
+      case Row(_: Vector, values: Seq[_]) =>
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/feature/BucketedRandomProjectionLSHSuite.scala]: So here do not need to select "keys" column ?

Comment (jkbradley) [reply mllib/src/test/scala/org/apache/spark/ml/feature/BucketedRandomProjectionLSHSuite.scala]: I don't think we have to.  The main thing here is to make sure that the transform really does happen.  Other tests check validity of the values.

Comment (WeichenXu123) [reply mllib/src/test/scala/org/apache/spark/ml/feature/BucketedRandomProjectionLSHSuite.scala]: OK. so I prefer to use simpler code:
```
testTransformer[Tuple1[Vector]](dataset.toDF(), brpModel, "values") {
    case Row(values: Seq[_]) =>
...
```

**Comment by WeichenXu123** [comment]
LGTM except a tiny issue. :)

**Comment by jkbradley** [comment]
Updated!  Thanks @WeichenXu123 -- I'll merge this once tests pass.

**Comment by jkbradley** [comment]
Merging with master

**Pull request closed without merging**